message_ix_models/tools/costs/projections.py

```diff
@@ -477,67 +489,115 @@ def create_message_outputs(
     )
 
     dtypes.update(year_act=int)
-    fom = (
-        df_merge.copy()
-        .drop(columns=["inv_cost"])
-        .assign(key=1)
-        .merge(
-            pd.DataFrame(data={"year_act": config.seq_years}).assign(key=1),
-            on="key",
-        )
-        .drop(columns=["key"])
-        .query("year_act >= year_vtg")
-        .assign(
-            val=lambda x: np.where(
-                x.year_vtg <= y_base,
-                np.where(
-                    x.year_act <= y_base,
-                    x.fix_cost,
+
+    if config.use_vintages is True:
+        fom = (
```

Comment:
As far as I can tell after squinting at it for a few minutes, the only difference between the two branches of the `if:` block is that one contains `.assign(val=lambda x…)`, and the other does not.

In this case, instead of duplicating a lot of code, I'd suggest to use a temporary/internal function that passes-through if nothing is to be done, something like:
```python

def _compute_value(df: pd.DataFrame) -> pd.Series:
    if config.use_vintages is False:
         return df["fix_cost"]  # Pass through

    return np.where(...)

fom = (
   df.merge.drop(columns=["inv_cost"])
   .assign(key=1)
   ...
   .assign(value=_compute_value, unit="USD/kWa")
   .rename(columns=…)
   ...
)
```

Reply:
Hi! Thanks all for both your comments. 

You're correct @khaeru that there's not a big difference between the two. However, the main differences are:

1. The `assign(val=lambda x...)` like you mentioned and
2. The `merge` command takes a different input. If `use_vintage` is True, then it is:
```
            .merge(
                pd.DataFrame(data={"year_act": config.seq_years}).assign(key=1),
                on="key",
            )
```
If it is False, then the merge command is:
```
            .merge(
                pd.DataFrame(data={"year_vtg": config.seq_years}).assign(key=1),
                on="key",
            )
```
So it is a matter of whether the dataframes are merged on `year_vtg` or `year_act`. Given this, would you still recommend implementing a function such as `_compute_value()`?

Reply:
Per (2), you can:
```python
# Construct a variable according the config setting:

to_merge = pd.DataFrame(
    {"year_act" if config.use_vintage else "year_vtg": config.seq_years}
).assign=key=1

# And then in the chained call, reference whatever the variable contains:
    ...
    .merge(to_merge, on="key")
    ...
```

Reply:
Sorry for the delay - thank you for your suggestions! I have edited the code to be a bit more condensed and not have to be duplicated as much:

https://github.com/iiasa/message-ix-models/blob/b3da5aa67c62c07f9dbe847a8ccc555d3b620f61/message_ix_models/tools/costs/projections.py#L489-L555

Reply:
Thanks—this makes the code easier to review and, in the future, easier to maintain.